src/Database/Esqueleto/Internal/JSON.hs

```diff
@@ -0,0 +1,196 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+module Database.Esqueleto.Internal.JSON
+    where
+
+import qualified Data.Aeson as Aeson
+import Data.Bifunctor (first)
+import qualified Data.ByteString.Lazy as LBS
+import Data.Proxy (Proxy(..))
+import Data.Text (Text)
+import qualified Data.Text as Text
+import Data.Text.Encoding (encodeUtf8)
+import qualified Data.Text.Lazy.Builder as TLB
+import Database.Esqueleto.Internal.Internal
+import Database.Persist
+
+class SqlToJson jsonValue a b | jsonValue a -> b where
+    toJson :: a -> SqlExpr (jsonValue b)
+
+class JsonAgg jsonValue where
+    jsonAgg :: SqlExpr (jsonValue a) -> SqlExpr (jsonValue [a])
+
+class JsonBuildArray jsonValue where
+    unsafeJsonbBuildArray :: UnsafeSqlFunctionArgument a => a -> SqlExpr (jsonValue b)
+
+class JsonBuildObject jsonValue where
+    unsafeJsonbBuildObject :: [(SqlExpr (Value Text), SqlExpr SomeValue)] -> SqlExpr (jsonValue a)
+
+multiset :: forall jsonValue a b r.
+            ( Aeson.FromJSON b
+            , SqlToJson jsonValue a b
+            , JsonAgg jsonValue
+            , SqlSelect (SqlExpr (jsonValue [b])) r
+            )
+         => SqlQuery a -> SqlExpr (jsonValue [b])
+multiset q =
+    subSelectUnsafe $ jsonAgg . toJson <$> q
+
+instance SqlToJson jsonValue (SqlExpr a) b
+      => SqlToJson jsonValue (SqlExpr (Maybe a)) (Maybe b) where
+    toJson =
+        let unMaybe :: SqlExpr (Maybe a) -> SqlExpr a
+            unMaybe = veryVeryUnsafeCoerceSqlExpr
+        in veryVeryUnsafeCoerceSqlExpr . toJson @jsonValue . unMaybe
+
+
+instance forall a jsonValue . (PersistEntity a, JsonBuildObject jsonValue)
+  => SqlToJson jsonValue (SqlExpr (Entity a)) (Entity a) where
+    toJson ent =
+        unsafeJsonbBuildObject fields
+        where
+            ed = entityDef $ Proxy @a
+            baseFields = fmap (\fieldDef ->
+                ( unsafeSqlValue $ TLB.fromText $ "'" <> unFieldNameHS (fieldHaskell fieldDef) <> "'"
```

Comment:
Does persistent support a way to read the to/from json field name given a field? 

We can sidestep the issue of to/from by using a custom decoder. As long as we are consistent at the boundaries we could even avoid the need to specify JSON instances for the types.

Reply:
Correct, the goal being to make something that automatically just works with the default JSON parser that gets generated.

Reply:
> which isn't going to be how the database wants to handle the JSON object.
@parsonsmatt I am not actually clear on what you mean by this? We aren't letting postgres do its default conversion.

Reply:
Yeah, I think I'm unsure that this is the right approach for converting something to a JSON representation.

I think I'd rather see something like:

```haskell
newtype JsonEntity a = JsonEntity (Entity a)

instance PersistEntity a => FromJSON (JsonEntity a) where
    parseJSON = withObject "JsonEntity" $ \o -> do
        let edef = entityDef (Proxy @a)
        ...
```

which definitely parses how a database would convert things to JSON. Then we aren't having to worry about custom encoders or decoders.


`persistent-2.14`'s `tabulateEntityA` could be useful here, but we may also need the `FieldDict` class from `prairie` to make it work how we want.

Reply:
So there isn't really a great way to do this in general without that `FieldDict` class since we need to know that the value on the other side has a `FromJSON` dictionary. 

Is there an example of someone using an exotic JSON parser? The thing that I need the default parser for is the underlying record not the `Entity record` since the `entityIdFromJSON` delegates to the `FromJSON` instance on the record.

Reply:
It should be noted that using the QQ's `json` tag will actually only allow you to configure the outer encoder/decoder but the record its self is always just `haskellName: someValue`